### ISSUE TITLE
Allow only valid group by SQL^2 queries

### ISSUE THREAD

**Issue by alissapajer** [quasar-analytics/quasar#2903]
SQL semantics allow for one of the following three entities to be selected in a query with a group by:

(1) the group key
(2) a reduction
(3) a constant value

For example
```sql
select count(pop), 42 as foo, state from zips group by state
```
is valid SQL while
```sql
select count(pop), 42 as foo, state, loc from zips group by state
```
is invalid SQL because `zips.loc` is not a group key (nor a reduction nor a constant).

Quasar should error in this case, giving a helpful message to the user. Currently this is allowed, and the query is evaluated by the connector.

EDIT: I originally stated we would implement this in the parser, but there is likely another better phase.

**Comment by jdegoes** [comment]
Why should we do this, since SQL2 has a sane semantic for it? It forbids useful classes of queries.

For example:

```sql
SELECT sales / SUM(sales) AS percent FROM orders
```

**Comment by jdegoes** [comment]
Also if we wanted to do this, I'd say the parser is probably not the best place for it, because it requires a dimensionality check. It would be better suited for an analysis phase on the parsed AST.

**Comment by alissapajer** [comment]
@jdegoes your example doesn't include an explicit grouping and so it wouldn't be affected by my proposal. 

What is the sane semantic for the following query (which would be disallowed under my proposal):
```sql
SELECT City, Country from Customers group by Country;
```
The only possibly sane semantic that I see is to arbitrarily choose a value for `City` for each unique `Country`. If that value is truly arbitrary, then the same query could return different results each time, even when run on the same connector, unless we require connectors to implement this arbitrary value consistently.

This query is disallowed in SQL because it "does not include the specified expression `City` as part of an aggregate function". So this would be an extension of SQL^2, which is not in and of itself a problem. I'm just not convinced this is an extension that makes sense.

**Comment by alissapajer** [comment]
(Also, I agree the parser may not be the correct place to do this.)

**Comment by jdegoes** [comment]
@alissapajer This query is the same as `SELECT City, Country from Customers`. In QScript, you build an extra layer of dimensionality, but then post `SELECT` that ends up being stripped off as only values are returned.

Another interesting example is, `SELECT City, SUM(Population) from Customers GROUP BY Country`, which has as many results as `Customers` and with columns with a constant population per country. This query may or may not do what you want, but it's still sensible.

**Comment by alissapajer** [comment]
@jdegoes hm, right, so considering your example
```sql
SELECT City, SUM(Population) from Customers GROUP BY Country
```
we use the group key to compute the reduction, and then return each value `City` paired with the population of its `Country`. This does appear to be a useful query, and avoids the problem I had with selecting an arbitrary value for `City`.

Still thinking about if I'm convinced this approach (or any other) is consistent.

**Comment by djspiewak** [comment]
@jdegoes It's valid to define semantics for those sorts of invalidly grouped queries, but those semantics are definitely incompatible with SQL.  Also, because of the semantics of SQL, you can generate queries which have non-deterministic results if you choose this path.

**Comment by alissapajer** [comment]
@jdegoes consider
```sql
SELECT Country from Customers group by Country
```
If the grouping information is ignored entirely when we compute the select, wouldn't this query return a set the size of `Customers` instead of a single row per unique value of `Country` (which is what I think it should return).

Furthermore, given the query
```sql
SELECT Country, SUM(Population) from Customers group by Country
```
I would expect a single row per `Country`. But if we are only incorporating the grouping information to compute the reduction, then we must return results the size of the original `Customers`, with multiple rows per `Country`, each paired with that `Country`'s population. To be clear, is this what you are proposing?

I don't see how selecting the group key (in this case `Country`) can in any way be "special" given your proposal. The grouping adds provenance information, but that's only used in the reduction. So do you think that selecting the group key should be treated that same as selecting any other non group key value?

**Comment by jdegoes** [comment]
An analogue would be `SELECT foo[_] FROM bar`. This query returns identical results to `SELECT foo[*] FROM bar`. Some constructs, including `GROUP BY`, add dimensional information that, if not utilized by subsequent reduction, is pretty useless. 

> I would expect a single row per Country. But if we are only incorporating the grouping information to compute the reduction, then we must return results the size of the original Customers, with multiple rows per Country, each paired with that Country's population. To be clear, is this what you are proposing?

If you utilize a naive translation into MRA, then yes, these are the semantics you get. Both `Country` and `SUM(Population)` are grouped by `Country`, so they auto-join on `Country` prior to the `OBJECT_CONCAT` that merges `{"Country": ...}` and `{"1": ...}`. Because there are as many `{"Country": ...}` entries as rows in the original data set, you would get results the size of the original `Customers`. 

**However**, the logical plan compiler inserts `ARBITRARY` around `Country` because it sees the same projection is used as a group by key and a selected projection. So the query you actually get is the following:

```sql
SELECT ARBITRARY(Country), SUM(Population) FROM Customers GROUP BY Country
```

In other words, you get SQL semantics and not "MRA semantics" in this case purely because our logical plan compiler detects the pattern and inserts the `ARBITRARY` there.

**Comment by alissapajer** [comment]
I don't think we should use MRA semantics for `group by` queries, because the results will not be what the user expects and the user will be very confused. This would be a huge change that would break many existing customers queries. If we want MRA grouping, I think we should introduce second alternative grouping notion with different syntax that represents MRA grouping semantics. I see the value in MRA grouping semantics, but I think we also need to provide regular `group by` semantics that are equivalent to SQL's.

**Comment by jdegoes** [comment]
@alissapajer I am saying we *already do* use MRA semantics for `GROUP BY` and that the user only gets what they expect because of a special-case in the logical plan compiler.

If that code isn't broken, then it should yield the same result for any class of query accepted by SQL.

**Comment by alissapajer** [comment]
@jdegoes :+1: I had misunderstood and thought that we were discussing removing the `Arbitrary` special case so that SQL^2 would redefine `group by` entirely, rather than extend `group by` like we do now. 

Let's keep this open for now as I (and maybe @djspiewak?) are still thinking about the details of the extension and are considering if there are any meaningful counterexamples.

**Comment by alissapajer** [comment]
@djspiewak I believe we have resolved this discussion as we've developed qsu, so I'm closing. Reopen if you disagree!
